### PR TITLE
#511 UniformItemsLayout fixes

### LINKED ISSUE #511: [Bug] Setting Margin on UniformItemsLayout crashes the app
### Description

Element not found. (0x80070490)

### Link to Reproduction Sample

https://github.com/VladislavAntonyuk/MauiSamples/tree/main/MauiPaint

### Steps to Reproduce

```xaml
<mct:UniformItemsLayout Margin="10" Padding="10">
						<Button BackgroundColor="Red" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="Green" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="Blue" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="Yellow" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="Gray" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="Black" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="White" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="Pink" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="Brown" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="Purple" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="SkyBlue" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
						<Button BackgroundColor="Orange" Command="{Binding SetLineColorCommand}" CommandParameter="{Binding Source={RelativeSource Self}, Path=BackgroundColor}"/>
					</mct:UniformItemsLayout>
```

### Expected Behavior

Margin/Padding applied

### Actual Behavior

Crash

### Basic Information

- Version with issue: 1.2.0

### PR DESCRIPTION
**Pull request by VladislavAntonyuk**
### Description of Change ###

- Fix UniformItemsLayout crashed on Margin set, 
- Padding not applied. 
- Fix Performance issue where we need only first item measure

 

 ### Linked Issues ###
 

 - Fixes #511

 ### PR Checklist ###
 
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/XYZ 

 ### Additional information ###

### CODE REVIEW

File: src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs [hunk 1]
```diff
@@ -104,7 +103,7 @@ int GetColumnsCount(int visibleChildrenCount, double widthConstraint)
 		=> Math.Min(
 			double.IsPositiveInfinity(widthConstraint)
 			   ? visibleChildrenCount
-			   : Math.Min((int)(widthConstraint / childWidth), visibleChildrenCount),
+			   : Math.Clamp((int)(widthConstraint / childWidth), 1, visibleChildrenCount),
```
Comment (bijington) [inline src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs:106]: I suspect the answer is no but is there any chance that `childWidth` can be `0` here?

**Review by bijington — APPROVED** [review]
Thanks @VladislavAntonyuk the changes look good. I just had one question on a bit of unchanged code

File: src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs (hunk 1 above)
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs:106]: theoretically yes, in practice no.
anyway, added some check to ensure we do not receive DevideByZeroException

File: src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs [hunk 2]
```diff
@@ -37,7 +37,7 @@ public int MaxRows
 	public int MaxColumns
 	{
 		get => (int)GetValue(MaxColumnsProperty);
-		set => SetValue(MaxColumnsProperty, value);
+		set => SetValue(MaxColumnsProperty, Math.Clamp(value, 1, int.MaxValue));
```
Comment (brminnick) [inline src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs:40]: I prefer we throw an exception when the user sets the value of `MaxColumns` below `1`.

This ensures that we educate the developer on the range of acceptable value.

If we use use `Math.Clamp`, then the user will try setting its value to `0` or `-1` and not see their desired behavior. They'll then try troubleshooting this bug, wasting their time, and may open a Bug Issue on our repo, wasting our time explaining that it's the expected behavior. 

Yes this exception will crash their app. But it is an Exception that the developer will only see once while testing their app and will quickly + easily fix after reading the `ArgumentOutOfRangeException` message.

```suggestion
		set => value >= 1 : SetValue(MaxRowsProperty, value) ? throw new new ArgumentOutOfRangeException(nameof(value), $"{nameof(MaxColumns)} must be greater or equal to {1}");
```

File: src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs [hunk 3]
```diff
@@ -47,28 +47,29 @@ public int MaxColumns
 	/// <returns>Child size</returns>
 	public Size ArrangeChildren(Rect rectangle)
 	{
-		Measure(rectangle.Width, rectangle.Height, MeasureFlags.None);
+		var width = Width - Padding.HorizontalThickness;
```
[the remark this answers is not in the document]
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs:50]: The quick answer is yes, but let me check it again

File: src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs [hunk 4]
```diff
@@ -37,7 +37,7 @@ public int MaxRows
 	public int MaxColumns
 	{
 		get => (int)GetValue(MaxColumnsProperty);
-		set => SetValue(MaxColumnsProperty, value);
+		set => value >= 1 : SetValue(MaxRowsProperty, value) ? throw new new ArgumentOutOfRangeException(nameof(value), $"{nameof(MaxColumns)} must be greater or equal to {1}");
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs:40]:
```suggestion
		set => value >= 1 : SetValue(MaxColumnsProperty, value) ? throw new new ArgumentOutOfRangeException(nameof(value), $"{nameof(MaxColumns)} must be greater or equal to {1}");
```

File: src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Layouts/UniformItemsLayout.shared.cs:66]: image

**Review by brminnick — APPROVED** [review]
Thanks Vlad!!

**Pull request merged**